### ISSUE TITLE
Set SO_REUSEADDR on the server sockets

### ISSUE THREAD

**Issue by Mosney** [AdguardTeam/AdGuardHome#1861]
### Prerequisites

Please answer the following questions for yourself before submitting an issue. **YOU MAY DELETE THE PREREQUISITES SECTION.**

- [x] I am running the latest version
- [x] I checked the documentation and found no answer
- [x] I checked to make sure that this issue has not already been filed

### Issue Details

* **Version of AdGuard Home server:**
  * AdGuardHome v0.102.0
* **How did you setup DNS configuration:**
  *  Router
* **If it's a router or IoT, please write device model:**
  *   amd64 device
* **Operating system and version:**
  *  OpenWrt 19.07.3

### Additional Information

I deploy a proxy software [clash](https://github.com/Dreamacro/clash) as gateway transparent proxy at OpenWrt which also run AdGuardHome, and AGH's config is `bind_host: 0.0.0.0`. Because of AGH deficiencies(see original issue below).   
In conclusion, AGH not allow reuse bind-address, so other processes cannot writeback via :53 port, cause proxy software behavioural peculiarities, it will lead proxied standard DNS query respond failed. And it will not affect local network DNS query which not through proxy at all. Of course, is a small probability bug, few users run AGH and proxy software on the same system. Nevertheless, I still hope you can check and fix it, AGH is a gateway level service, it should robustness and compatibility enough.

Original issue: https://github.com/Dreamacro/clash/issues/616#issuecomment-611961492
Comment translation:
Some programs bind to` 0.0.0.0:53` or `:::53` and no set SO_REUSEADDR, which makes it impossible to bind the WriteBack By 8.8.8.8:53

**Comment by Mosney** [comment]
Forgot to mention it: Don't use 0.0.0.0/:: to bind is a quick way to fix this issue, but the drawbacks are obvious, you will no longer be able to listen all IP.

**Comment by ameshkov** [comment]
This `SO_REUSEADDR` is an old golang issue, but it seems I've finally found a way to fix it.

@szolin check this out: https://gist.github.com/joliver/4ccd58605e07e8edf71904b172d95513

I am not sure if this approach will work properly on Windows, but other than that, it seems okay.

**Comment by szolin** [comment]
I don't understand how it's possible.

> Some programs bind to 0.0.0.0:53 or :::53 and no set SO_REUSEADDR, which makes it impossible to bind the WriteBack By 8.8.8.8:53

SO_REUSEADDR won't help in case AGH is listening on 0.0.0.0

> check this out: https://gist.github.com/joliver/4ccd58605e07e8edf71904b172d95513

SO_REUSEPORT is a feature for distributing network load between several processes of the same application.  If we start using it, we will receive half of the incoming requests.  The other half will be received by another application.

**Comment by ihipop** [comment]
> I don't understand how it's possible.
> 
> > Some programs bind to 0.0.0.0:53 or :::53 and no set SO_REUSEADDR, which makes it impossible to bind the WriteBack By 8.8.8.8:53

It's possible. 
If you dive into the Linux TPROXY stack,when processing a UDP connection, for example ,if the Dest Addr is `8.8.8.8:53`, the TPROXY program must bind to `8.8.8.8:53` on TPROXY host os and send the UDP packet back to the Client to make sure the Src Addr and Port is `8.8.8.8:53` , IF,when ADG has binded :53 on this host os without  `SO_REUSEADDR`, the behavior of the TPROXY program will fail with `EADDRINUSE (Address already in use)`

Here are some example of TPROXY program that has this behavior

https://github.com/Dreamacro/clash/blob/8c7c8f43741923d6a7650dc723b4b51b0f540777/listener/tproxy/packet.go#L20

https://github.com/v2fly/v2ray-core/blob/541fb5c5b173cc60b2009fa0fe3ca4b27133f40c/proxy/dokodemo/dokodemo.go#L161-L168

> SO_REUSEADDR won't help in case AGH is listening on 0.0.0.0
> 
> > check this out: https://gist.github.com/joliver/4ccd58605e07e8edf71904b172d95513
> 
> SO_REUSEPORT is a feature for distributing network load between several processes of the same application. If we start using it, we will receive half of the incoming requests. The other half will be received by another application.

it's true when working with `distributing network load between several processes`,
but in the scene above I mentioned ,it's really safe for ADG to use `SO_REUSEADDR`,because we wont get in-bond connection to `8.8.8.8:53` on our HOST
And,`DNSMASQ` also allow `SO_REUSEADDR/SO_REUSEPORT` by default


IMO, we shoule make this behavior configurable... 


@ameshkov @szolin

**Comment by ihipop** [comment]
after 5 days since this comment,the Team seems to take some actions to support `SO_REUSEADDR`

https://github.com/AdguardTeam/dnsproxy/commit/5864a879fd36361c644d99f94af14aae9b7ed9b6#diff-8a752bfb9f71b4905540ffeaf78349f38982a8283eac10f845f3eaeb58ba63f6R19

I just can't understand why `SO_REUSEPORT` is also used. let's wait and see
